# Model 1: Revisit of Titanic Dataset

### Remember that the logistic regression is a classifier as well. Name this first model model1. Can you use lines 1-23 under Part 3 of W7 to run a logistic regression with gridsearchCV to predict death on the titanic.

In [2]:
import pandas as pd
import sklearn
import numpy as np
from sklearn.preprocessing import MinMaxScaler 
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split 
df=pd.read_csv('The Titanic Dataset.csv', header=1)
from sklearn.linear_model import LogisticRegression

### copy over data cleaning and normalizing from W6

In [3]:
df2=df.drop(columns=['embarked'])
df2=df2.replace('?',np.nan) #replace the ? string with not a number 
df2['gender']=df2['gender'].replace({'female':0,'male':1}) #convert gender to digits
df2['gender']=pd.to_numeric(df2['gender'],errors='coerce') #replaces unknown values with NaN not throw an error
df2['gender']=df2['gender'].fillna(df2['gender'].mode()[0]) #supposedly replaces the NaN in row 4 with the most common gender but I literally can't figure it out
df2['age']=pd.to_numeric(df2['age'],errors='coerce') #replaces unknown values with NaN not throw an error
df2['age']=df2['age'].fillna(df2['age'].median()) #convert all NaN to median (convention)
df2['fare']=pd.to_numeric(df2['fare'],errors='coerce') #replaces unknown values with NaN not throw an error
df2['fare']=df2['fare'].fillna(df2['fare'].median())
df2['family']=df2['family'].fillna(df2['family'].median())
df2.head(10)

,sn,pclass,survived,Unnamed: 3,gender,age,family,fare,date
0,1,3,0,Mr. Anthony,1.0,42.0,0.0,7.5500,01-Jan-90
1,1,3,0,Mr. Anthony,1.0,42.0,0.0,7.5500,01-Jan-90
2,2,3,0,Master. Eugene Joseph,1.0,28.0,2.0,20.2500,02-Jan-90
3,3,2,0,"Abbott, Mr. Rossmore Edward",1.0,28.0,2.0,14.4542,03-Jan-90
4,4,3,1,"Abbott, Mr. Rossmore Edward",0.0,35.0,2.0,20.2500,04-Jan-90
5,5,3,1,"Abelseth, Miss. Karen Marie",0.0,16.0,0.0,7.6500,05-Jan-90
6,6,3,1,"Abelseth, Mr. Olaus Jorgensen",1.0,25.0,0.0,7.6500,06-Jan-90
7,7,2,0,"Abelson, Mr. Samuel",1.0,30.0,1.0,24.0000,07-Jan-90
8,8,2,1,"Abelson, Mrs. Samuel (Hannah Wizosky)",0.0,28.0,1.0,24.0000,08-Jan-90
9,9,3,1,"Abrahamsson, Mr. Abraham August Johannes",1.0,20.0,0.0,7.9250,09-Jan-90


### Split data

In [4]:
X=df2.drop(columns=['Unnamed: 3','date','survived'])
y=df2['survived']
Xtrain,Xtest,ytrain,ytest=train_test_split(X,y,test_size=0.2,random_state=67)

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
log_reg = LogisticRegression()

param_grid = {
'penalty': ['l1', 'l2'],
'C': [0.001, 0.01, 0.1, 1, 10, 100],
'solver': ['liblinear'] # 'liblinear' works well for small/medium datas
}
grid_log = GridSearchCV(log_reg,param_grid,cv=5,scoring='accuracy')
grid_log.fit(Xtrain, ytrain)
print(f"Best Penalty: {grid_log.best_params_['penalty']}")
print(f"Best C: {grid_log.best_params_['C']}")
best_log_model = grid_log.best_estimator_

/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1160: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=

Best Penalty: l1
Best C: 1


/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1160: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=

## Parameters
Best penalty: L1
- this means the best penalty is Lasso; some coefficients, usually the useless ones, are shrunken to exactly 0 (or essentially killed off) 
- C=1 is the default for regularization, meaning an adequate/medium amount of regrularization is best for my model

## Use appropriate metrics for logistic regression to see how your model did.
accuracy, AUC/ROC, and others you may have learned (precision/recall). Make a choice
regarding which metric is the most appropriate for death prediction, justify your choice in metric.
Analyze your results.

In [6]:
model=best_log_model.fit(Xtrain,ytrain)
ypredict=model.predict(Xtest)
accuracy=accuracy_score(ytest, ypredict)
print(f"Accuracy: {accuracy:.4f}")
print(f"Auc score: {metrics.roc_auc_score(ytest, ypredict):.4f}")
print(f"Recall: {metrics.recall_score(ytest, ypredict):.4f}")
print(f"F1 Score: {metrics.f1_score(ytest, ypredict):.4f}")


Accuracy: 0.8008
Auc score: 0.7760
Recall: 0.6700
F1 Score: 0.7204


/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:1160: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(


## Analysis of data 
accuracy score: 80% correct at identifying either a survivor or a dead person--> overall model is pretty good at classifying the passengers as a whole

Auc score: model correctly identifies survivor over dead person as more likely to survive 78% of the time

recall: out of all survivors model only correctly identified 67% of them 

f1 score: pretty high, means model balances recall and precision well 


## Model Selection explanation 
Given what we are trying to predict (death or survival), do you think it is appropriate to use a
linear regression instead of a logistic one? Why or why not?

It is not appropriate to use a linear regression instead of a logistic one because there are only two categories, survived or dead, there is no in between (no partially dead, nobody is more dead than another it isn't a spectrum) and since logistic regression classifies into such categories, it is more appropriate to use a logistic regression.

# Model 2: Revisit of the Spotify Dataset

## Data selection 
We used linear regression to predict danceability the first time around on W4. Take a look at your
columns again. Make a meaningful choice of 3 columns that you think may predict danceability a
bit better this time for when you split your data

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn 
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score, recall_score, f1_score
df=pd.read_csv("top2019.csv")
scaler=MinMaxScaler()
non_numeric=['id','name','artists']
numeric=df.columns.difference(non_numeric)
scaler=MinMaxScaler()
df[numeric]=scaler.fit_transform(df[numeric])
ndf=df[numeric]

## Use part 5 on W7 to run a regression that allows use of gridsearchCV to run a lasso regression on danceability. What were your best parameters? Interpret your results.

In [13]:
X=ndf[['valence','speechiness','acousticness']]
y=ndf['danceability']
Xtrain,Xtest,ytrain,ytest=train_test_split(X,y,test_size=0.2,random_state=67)
param_grid = {
'alpha': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100], 
}
from sklearn.linear_model import Lasso
lasso = Lasso()
Lassomodel= GridSearchCV(lasso, param_grid, cv=5, scoring='r2')
Lassomodel.fit(Xtrain, ytrain)
print(f"Best Alpha: {Lassomodel.best_params_['alpha']}")
best_lasso_model = Lassomodel.best_estimator_
model=best_lasso_model.fit(Xtrain,ytrain)
ypredict=model.predict(Xtest)
print(f"R2 Score: {metrics.r2_score(ytest, ypredict):.4f}")


Best Alpha: 0.001
R2 Score: 0.2857


Best parameters were alpha=0.001 (less regularization) and best r2 is 0.2857 meaning that my regression shows a relatively small relationship between danceability and my three columns

## Use part 5 on W7 to run a regression that allows use of gridsearchCV to run a ridge regression on danceability. What were your best parameters? Interpret your results

In [14]:
X=ndf[['valence','speechiness','acousticness']]
y=ndf['danceability']
Xtrain,Xtest,ytrain,ytest=train_test_split(X,y,test_size=0.2,random_state=67)
param_grid = {
'alpha': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100], 
}
from sklearn.linear_model import Ridge
ridge = Ridge()
Ridgemodel= GridSearchCV(ridge, param_grid, cv=5, scoring='r2')
Ridgemodel.fit(Xtrain, ytrain)
print(f"Best Alpha: {Ridgemodel.best_params_['alpha']}")
best_ridge_model = Ridgemodel.best_estimator_
model=best_ridge_model.fit(Xtrain,ytrain)
ypredict=model.predict(Xtest)
print(f"R2 Score: {metrics.r2_score(ytest, ypredict):.4f}")


Best Alpha: 1
R2 Score: 0.3324


Best parameters were alpha=1 (medium/default regularization) and best r2 is 0.0334 meaning that my regression shows a relatively small relationship, but bigger than the lasso one, between danceability and my three columns